### PR TITLE
Implement transaction coordinator

### PR DESCRIPTION
**Pull request by pattyshack**
This refactors block database and result collector into transaction coordinator. This also adds synchronization functionality to be used for parallel execution.

### CODE REVIEW

File: engine/execution/computation/computer/transaction_coordinator.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	startErr := coordinator.abortErr
+	for {
+		if coordinator.snapshotTime > snapshotTime ||
+			coordinator.abortErr != nil {
+
+			return startTime, startErr, coordinator.snapshotTime, coordinator.abortErr
+		}
+
+		coordinator.cond.Wait()
+	}
```
Comment (janezpodhostnik) [inline engine/execution/computation/computer/transaction_coordinator.go:183]:
```suggestion
	for coordinator.snapshotTime <= snapshotTime && coordinator.abortErr != nil  {
            coordinator.cond.Wait()
	}
		
        return startTime, startErr, coordinator.snapshotTime, coordinator.abortErr
```
maybe a bit simpler. Just a suggestion no strong feelings about it.

**Review by janezpodhostnik — APPROVED** [review]

File: engine/execution/computation/computer/transaction_coordinator.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	startErr := coordinator.abortErr
+	for coordinator.snapshotTime <= snapshotTime && coordinator.abortErr == nil {
+		coordinator.cond.Wait()
+	}
+
+	return startTime, startErr, coordinator.snapshotTime, coordinator.abortErr
+}
+
+func (txn *transaction) WaitForUpdates() error {
+	_, _, _, err := txn.coordinator.waitForUpdatesNewerThan(txn.SnapshotTime())
```
Comment (zhangchiqing) [inline engine/execution/computation/computer/transaction_coordinator.go:183]: why does `waitForUpdatesNewerThan` need to return two pairs of (value, error), why the `startErr` is being ignored?

File: engine/execution/computation/computer/transaction_coordinator.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		return err
+	}
+
+	coordinator.writeBehindLog.AddTransactionResult(
+		txn.request,
+		executionSnapshot,
+		txn.Output(),
+		time.Since(txn.startedAt))
+
+	coordinator.snapshotTime += 1
```
Comment (zhangchiqing) [inline engine/execution/computation/computer/transaction_coordinator.go:152]:
```suggestion
	// could you add some comment to explain why to increment the snapshot time
	coordinator.snapshotTime += 1
```

**Review by zhangchiqing — APPROVED** [review]

**Comment by pattyshack** [comment]
bors merge

**Comment by pattyshack** [comment]
bors merge

**Pull request merged**